engine/access/rpc/backend/connection_factory.go

```diff
@@ -93,55 +105,44 @@ func (cf *ConnectionFactoryImpl) createConnection(address string, timeout time.D
 		grpc.WithDefaultCallOptions(grpc.MaxCallRecvMsgSize(int(cf.MaxMsgSize))),
 		grpc.WithTransportCredentials(insecure.NewCredentials()),
 		grpc.WithKeepaliveParams(keepaliveParams),
-		WithClientUnaryInterceptor(timeout))
+		grpc.WithChainUnaryInterceptor(connInterceptors...),
+	)
 	if err != nil {
 		return nil, fmt.Errorf("failed to connect to address %s: %w", address, err)
 	}
 	return conn, nil
 }
 
 func (cf *ConnectionFactoryImpl) retrieveConnection(grpcAddress string, timeout time.Duration) (*grpc.ClientConn, error) {
-	var conn *grpc.ClientConn
-	var store *CachedClient
-	cacheHit := false
 	cf.mutex.Lock()
```

Comment:
This is a global lock, which is held through the whole connection process. This could block for seconds in the worst case.

It was originally used to make the `Get`/`Add` operation atomic, and there was a per client mutex to allow releasing the global lock during connections.

One option to improve this is to use the `PeekOrAdd` method to atomically check for an existing item or add a new one if none exist. If you do that, I _think_ the global mutex could be removed, and you could add back the per client mutex to protect the critical section between where it checks if `store.ClientConn` is nil, and otherwise create the connection. That way, any other goroutines that are want a connection will block until the connection is setup and then use the new connection.

alternatively, you could keep the 2 locks.

Reply:
looks like the call to `Get` was removed from `retrieveConnection`, which results in clients never being reused. Can you add it back in?

Reply:
ah, I see the confusion.

What I meant in that comment was that when we need to recreate a connection to a cached client because it was invalidated or evicted, we should ignore the original struct and create a new one in `retrieveConnection`. This is to ensure that any existing goroutines that still depend on that specific instance will continue to work and gracefully shutdown the connection.

For other cases, we still need to get the cached client, otherwise there's not much sense in caching them in the first place. I think the original logic is mostly fine, though there may be some race conditions to consider between the call to remove and checking in the `retrieveConnection` function.